# Laboratório 1 – Arrays e DataFrames [<img src="https://raw.githubusercontent.com/urielmoreirasilva/ICX532/main/labs_gabs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29/images/colag_logo.svg" style="float: right; margin-right: 0%; vertical-align: middle; width: 6.5%;">](https://colab.research.google.com/github/urielmoreirasilva/ICX532/blob/main/labs_gabs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29.ipynb) [<img src="https://raw.githubusercontent.com/urielmoreirasilva/ICX532/main/labs_gabs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29/images/github_logo.svg" style="float: right; margin-right: 0%; vertical-align: middle; width: 3.25%;">](https://github.com/urielmoreirasilva/ICX532/blob/main/labs_gabs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29.ipynb)

Bem-vindo ao Laboratório 1! Nesta tarefa, aprenderemos mais sobre as propriedades e técnicas de manipulação de arrays e DataFrames.

Você deve concluir todo este laboratório e enviá-lo ao Moodle até às 23h59 da data de vencimento.

### Referências
- [BPD: Capítulos 7 a 9](https://notes.dsc10.com/)
- [CIT: Capítulos 5 e 6](https://inferentialthinking.com/)
- Aulas: Tópicos 1, 2 e 3. 

Material adaptado do [DSC10 (UCSD)](https://dsc10.com/) por [Flavio Figueiredo (DCC-UFMG)](https://flaviovdf.io/fcd/) e [Uriel Silva (DEST-UFMG)](https://urielmoreirasilva.github.io)

**<u> NOTA</u>: Não use loops `for` em nenhuma pergunta deste laboratório.** Lembre do Tópico 08 que os loops em Python são lentos, e os loops em `arrays` e `DataFrame`s geralmente devem ser evitados. Como alternativa, devemos idealmente utilizar as funções nativas de bibliotecas como `numpy` e `pandas` para a manipulação desses objetos.

In [1]:
## Imports para esse laboratório
import numpy as np
import pandas as pd

## Opções do MatplotLib 
import matplotlib.pyplot as plt
plt.style.use('ggplot')
plt.rcParams['figure.figsize'] = (10, 5)

## 1. Arrays

## 1.1 Primeiras Definições

#### **Pergunta 1.1.**

Defina um `array` contendo os números 2, 4 e 6, nessa ordem. Nomeie esse array como `numeros_pares`.

In [3]:
numeros_pares = [2, 4, 6]
numeros_pares

[2, 4, 6]

#### **Pergunta 1.2.** 

Crie um `array` contendo os números 0, -1, 1, $\pi$ e $e$, nessa ordem. Nomeie esse array como `outros_numeros`.

*<u> Dica</u>: $\pi$ e $e$ estão disponíveis no módulo `np`, que já foi importado. Uma alternativa aqui seria utilizar `math.pi` para obter $\pi$, mas prefira utilizar `np.pi` aqui, e evite carregar uma biblioteca (isto é, `math`) a mais desnecessariamente. Utilize um raciocínio análogo para $e$.*

In [4]:
outros_numeros = [0, -1, 1, np.pi, np.e]
outros_numeros

[0, -1, 1, 3.141592653589793, 2.718281828459045]

#### **Pergunta 1.3.** 

Crie um `array` contendo cinco strings: `"Hello"`, `","`, `" "`, `"world"` e `"!"` (note que o terceiro elemento aqui é simplesmente um espaço único entre aspas). Nomeie esse array como `componentes_hello_world`.

*<u> Nota</u>: Dependendo do seu compilador, IDE e plataforma, ao imprimir `componentes_hello_world`, você pode notar algumas informações extras, como por exemplo `dtype='<U5'`. Essa é apenas a maneira do NumPy informar que os elementos do array são strings. Caso você esteja interessado em saber mais, o `U` significa que esta string está codificada em [unicode](https://en.wikipedia.org/wiki/Unicode), e o `<5` significa que todas as strings no array têm 5 caracteres ou menos.*

In [10]:
componentes_hello_world = ["Hello", ",", " ", "world", "!"]
componentes_hello_world

['Hello', ',', ' ', 'world', '!']

Muitas vezes, em Ciência de Dados, queremos trabalhar com muitos números espaçados uniformemente dentro de algum intervalo. A biblioteca NumPy fornece uma função especial para isso, chamada `arange`. A expressão `np.arange(comeco, fim, espaco)` produz um array com todos os números começando em `comeco`, contando de  `espaco` em `espaco`, e parando logo **antes** de `fim` ser alcançado$^1$.

Por exemplo, o valor de `np.arange(1, 8, 2)` é um array com os elementos `1, 3, 5 e 7`, isto é, começando em 1 e contando de 2 em 2, terminando até chegar no último valor menor que 8 (que é 7). Em outras palavras, essa chamada cria o mesmo array que `np.array([1, 3, 5, 7])`.

$^1$*<u> Nota</u>: Esse comportamento é consistente com a indexação do Python, que se inicia em `0`*.

##### **Pergunta 1.1.4.** 

Use `np.arange` para criar um array com todos os múltiplos de 99 de 0 até (**e incluindo**) 9999. Portanto, seus elementos serão 0, 99, 198, 297, etc.

In [16]:
## Resposta usual (calcula-se antes 9999 + 99 = 10098, e coloca-se esse número no final)
multiplos_de_99 = np.arange(0, 10098, 99)
multiplos_de_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [18]:
## Resposta alternativa #1 (mais simples)
multiplos_de_99 = np.arange(0, 9999 + 99, 99)
multiplos_de_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

In [19]:
## Resposta alternativa #2 (mais comum na prática)
multiplos_de_99 = np.arange(0, 9999 + 1, 99)
multiplos_de_99

array([   0,   99,  198,  297,  396,  495,  594,  693,  792,  891,  990,
       1089, 1188, 1287, 1386, 1485, 1584, 1683, 1782, 1881, 1980, 2079,
       2178, 2277, 2376, 2475, 2574, 2673, 2772, 2871, 2970, 3069, 3168,
       3267, 3366, 3465, 3564, 3663, 3762, 3861, 3960, 4059, 4158, 4257,
       4356, 4455, 4554, 4653, 4752, 4851, 4950, 5049, 5148, 5247, 5346,
       5445, 5544, 5643, 5742, 5841, 5940, 6039, 6138, 6237, 6336, 6435,
       6534, 6633, 6732, 6831, 6930, 7029, 7128, 7227, 7326, 7425, 7524,
       7623, 7722, 7821, 7920, 8019, 8118, 8217, 8316, 8415, 8514, 8613,
       8712, 8811, 8910, 9009, 9108, 9207, 9306, 9405, 9504, 9603, 9702,
       9801, 9900, 9999])

### Leituras de temperatura 🌡️

A NOAA (Administração Nacional Oceânica e Atmosférica dos EUA) opera estações meteorológicas que medem as temperaturas da superfície em diferentes locais dos Estados Unidos. As leituras horárias estão [disponíveis publicamente](http://www.ncdc.noaa.gov/qclcd/QCLCD?prior=N).

Suponha que baixemos todos os dados de horários do site de San Diego, Califórnia, para o mês de dezembro de 2021. Para analisar os dados, queremos saber quando cada leitura foi feita, mas descobrimos que os dados não incluem as datas nem o horário das leituras.

No entanto, sabendo que a primeira leitura foi feita no primeiro instante de dezembro de 2021 (meia-noite do dia 1º de dezembro) e que cada leitura subsequente foi feita exatamente 1 hora após a última, podemos criar uma "grade" de datas e horários para utilizar como referência.

#### **Pergunta 1.5.** 

Crie um `array` com o tempo, *em segundos*, desde o início do mês (isto é, meia-noite do dia 1º de dezembro) em que cada leitura foi feita. Nomeie esse array comocomo `registro_de_leituras`.

*<u>Dica #1</u>: Dezembro tem 31 dias, o que equivale a ($31 \times 24$) horas, ou ($31 \times 24 \times 60 \times 60$) segundos.*

*<u>Dica #2</u>: A função `len` também funciona em arrays. Para checar seus resultados, verifique se seu array `registro_de_leituras` tenha $31 \times 24$ elementos, já que as leituras são feitas de hora em hora durante 31 dias.*

In [21]:
## Uma solução possível: cada hora tem 3600 segundos
registro_de_leituras = np.arange(0, 31 * 24 * 3600, 3600)
registro_de_leituras

array([      0,    3600,    7200,   10800,   14400,   18000,   21600,
         25200,   28800,   32400,   36000,   39600,   43200,   46800,
         50400,   54000,   57600,   61200,   64800,   68400,   72000,
         75600,   79200,   82800,   86400,   90000,   93600,   97200,
        100800,  104400,  108000,  111600,  115200,  118800,  122400,
        126000,  129600,  133200,  136800,  140400,  144000,  147600,
        151200,  154800,  158400,  162000,  165600,  169200,  172800,
        176400,  180000,  183600,  187200,  190800,  194400,  198000,
        201600,  205200,  208800,  212400,  216000,  219600,  223200,
        226800,  230400,  234000,  237600,  241200,  244800,  248400,
        252000,  255600,  259200,  262800,  266400,  270000,  273600,
        277200,  280800,  284400,  288000,  291600,  295200,  298800,
        302400,  306000,  309600,  313200,  316800,  320400,  324000,
        327600,  331200,  334800,  338400,  342000,  345600,  349200,
        352800,  356

In [22]:
## checando os resultados com a função len (31 * 24 = 744)
len(registro_de_leituras)

744

## 1.2. Indexação

Vamos agora voltar a nossa atenção para um conjunto de dados um pouco mais interessante. A célula de código abaixo cria um array denominado `populacao`, que inclui populações mundiais estimadas em cada ano de **1950** a **2022**. (As estimativas vêm da seguinte [base de dados internacional](https://www.census.gov/data-tools/demo/idb/#/country?COUNTRY_YEAR=2022&COUNTRY_YR_ANIM=2022), mantida pelo US Census Bureau, órgão equivalente ao IBGE do Brasil).

Em vez de digitar os dados manualmente, nós os carregamos de um arquivo armazenado na nuvem chamado `world_population_2022.csv`.

In [28]:
## Veremos em detalhes as particularidades desse tipo de chamada mais adiante;
## -- por enquanto, apenas ignore os detalhes específicos e execute a célula
populacao = pd.read_csv("https://raw.githubusercontent.com/urielmoreirasilva/ICX532/main/labs_gabs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames%20%28gabarito%29/data/world_population_2022.csv").get("Population").values
populacao

array([2557619597, 2594942227, 2636777090, 2682060684, 2730237675,
       2782111389, 2835315327, 2891368627, 2948159570, 3000742521,
       3043031253, 3084053711, 3140239653, 3210037409, 3281477826,
       3350773176, 3421097064, 3490825940, 3562887008, 3637819236,
       3713457589, 3791172327, 3867519813, 3943132388, 4017779234,
       4089387557, 4159536915, 4230430893, 4301282222, 4374940345,
       4445975606, 4527418598, 4610620221, 4694937687, 4777055423,
       4862317393, 4949951891, 5040273543, 5131575729, 5222662682,
       5315511894, 5403253915, 5490481497, 5568231516, 5650178207,
       5733211108, 5815333785, 5895837672, 5975189305, 6053955779,
       6132455985, 6211328357, 6290282107, 6369186797, 6448262425,
       6527056809, 6607396274, 6689442159, 6773319540, 6857160919,
       6939761510, 7022084781, 7105001721, 7188528811, 7271598780,
       7353476064, 7435151387, 7516769535, 7597066210, 7676686052,
       7756873419, 7831718605, 7905336896])

Veja abaixo como obtemos o primeiro elemento de `populacao`, que é a população mundial no primeiro ano do conjunto de dados, 1950.

In [27]:
populacao[0]

np.int64(2557619597)

Observe atentamente que na expressão acima utilizamos colchetes `[]`. Os colchetes sinalizam que estamos *acessando* um elemento do array. Colchetes em Python são como os subscritos que utilizamos em matemática para denotar os elementos de uma sequência (por exemplo $x_1$, $x_2$, ...).

O valor da expressão acima é o número $2557619597$ (cerca de 2,5 bilhões), porque é o primeiro elemento do array `populacao`.

É importante notar que aqui escrevemos `populacao[0]`, e não `populacao[1]`, para obter o primeiro elemento. Esta é uma convenção do Python, em que `0` é chamado de *índice* do primeiro item. Seguindo essa lógica, então `1` seria o índice do segundo item, `2` seria o índice do terceiro, e assim por diante.

Abaixo seguem mais alguns exemplos de indexação.

Nesses exemplos, demos nomes às saídas que obtemos como elementos de `populacao`.

In [29]:
## O terceiro elemento do array é a população em 1952
populacao_1952 = populacao[2]
populacao_1952

np.int64(2636777090)

In [30]:
## O décimo terceiro elemento do array é a população em 1962 (que é 1950 + 12)
populacao_1962 = populacao[12]
populacao_1962

np.int64(3140239653)

In [31]:
## O 73º elemento do array é a população em 2022
populacao_2022 = populacao[72]
populacao_2022

np.int64(7905336896)

Como o array `populacao` possui 73 elementos, é impossível acessar um índice maior do que `72`. Note que também é impossível acessar elementos que não sejam números inteiros. 

In [37]:
# ## Descomente e execute!
# populacao_2023 = populacao[73]
# populacao_2023

In [38]:
# ## Descomente e execute!
# populacao[1.50]

Voltaremos a esse ponto novamente adiante, mas tentar acessar índices negativos (por exemplo `-1`) talvez à primeira vista não reproduza o efeito originalmente esperado.

In [40]:
populacao

array([2557619597, 2594942227, 2636777090, 2682060684, 2730237675,
       2782111389, 2835315327, 2891368627, 2948159570, 3000742521,
       3043031253, 3084053711, 3140239653, 3210037409, 3281477826,
       3350773176, 3421097064, 3490825940, 3562887008, 3637819236,
       3713457589, 3791172327, 3867519813, 3943132388, 4017779234,
       4089387557, 4159536915, 4230430893, 4301282222, 4374940345,
       4445975606, 4527418598, 4610620221, 4694937687, 4777055423,
       4862317393, 4949951891, 5040273543, 5131575729, 5222662682,
       5315511894, 5403253915, 5490481497, 5568231516, 5650178207,
       5733211108, 5815333785, 5895837672, 5975189305, 6053955779,
       6132455985, 6211328357, 6290282107, 6369186797, 6448262425,
       6527056809, 6607396274, 6689442159, 6773319540, 6857160919,
       6939761510, 7022084781, 7105001721, 7188528811, 7271598780,
       7353476064, 7435151387, 7516769535, 7597066210, 7676686052,
       7756873419, 7831718605, 7905336896])

In [41]:
## quando o índice é negativo, contamos na ordem inversa!
## -- dessa forma, o índice [-1] é na verdade o **último** elemento do array
populacao[-1]

np.int64(7905336896)

#### **Pergunta 1.6.** 

Defina abaixo a variável `populacao_1998` para receber a população mundial em 1998. Faça essa declaração obtendo o elemento apropriado de `populacao`.

In [42]:
populacao_1998 = populacao[37]
populacao_1998

np.int64(5040273543)

## 1.3. Iterando sobre os elementos de um array

Arrays são muito úteis para realizar uma mesma operação muitas vezes. Com o uso de arrays, podemos acessar (e operar) em múltiplos elementos de uma vez só!

### Logaritmos
Aqui está uma pergunta simples que podemos fazer sobre a população mundial:

> Qual era o tamanho da população em *ordens de magnitude* em cada ano?

A função logaritmo é uma forma de medir o tamanho de um número. O logaritmo (base 10) de um número aumenta em 1 cada vez que multiplicamos o número por 10. É como uma medida de quantos dígitos decimais o número possui ou quão grande ele é em ordens de grandeza.

Poderíamos tentar responder nossa pergunta assim, usando a função `log10` do NumPy em cada elemento do array `populacao`:

In [43]:
magnitude_da_populacao_1950 = np.log10(populacao[0])
magnitude_da_populacao_1951 = np.log10(populacao[1])
magnitude_da_populacao_1952 = np.log10(populacao[2])
magnitude_da_populacao_1953 = np.log10(populacao[3])

# ... e assim por diante

Mas isso é tedioso e repetitivo. Deve haver uma maneira melhor!

Acontece que o `log10` do NumPy é bastante poderoso. Ele não apenas pode receber um único número (como `populacao[0]`) como entrada e retornar o logaritmo de um único número, mas **também** pode receber um array de números e retornar o logaritmo de cada elemento desse array!

Se você fornecer ao `log10` do NumPy um array como entrada, ele retornará um array do mesmo tamanho, onde o primeiro elemento do resultado é o logaritmo do primeiro elemento da entrada, o segundo elemento do resultado é o logaritmo de o segundo elemento da entrada e assim por diante.

<img src="https://raw.githubusercontent.com/urielmoreirasilva/ICX532/main/labs/Laborat%C3%B3rio%201%20%E2%80%93%20Arrays%20e%20DataFrames/images/array_logarithm.jpg">

Isso é chamado de aplicação *elementwise* (elemento a elemento) da função, uma vez que opera separadamente em cada elemento do array em que é chamada.

**Pergunta 1.3.1.** Use a função `log10` do NumPy para calcular os logaritmos da população mundial em cada ano. Dê ao resultado (uma matriz de 73 números) o nome `magnitude_da_populacao`. Seu código deve ser muito curto.

In [ ]:
magnitude_da_populacao = ...
magnitude_da_populacao

##### Aritmética
A aritmética também funciona *elementwise* em arrays. Por exemplo, você pode dividir todos os números da população por 1 bilhão para obter números em bilhões:

In [ ]:
populacao_em_bilhoes = populacao / 1000000000
populacao_em_bilhoes

Você pode fazer o mesmo com adição, subtração, multiplicação e exponenciação (`**`). Por exemplo, você pode calcular uma gorjeta de vinte por cento em várias contas de restaurante de uma só vez:

In [ ]:
contas_dos_restaurantes = np.array([20.12, 39.90, 31.01])
print("Conta dos restaurantes:\t", contas_dos_restaurantes)

gorjetas = 0.2 * contas_dos_restaurantes
print("Gorjetas:\t\t", gorjetas)

<img src="https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs_gabs/lab01/imagens/array_multiplication.jpg?raw=true">

**Pergunta 1.3.2.** Suponha que a cobrança total em um restaurante seja a conta original mais a gorjeta (20%). Isso significa que podemos multiplicar a fatura original por 1.2 para obter a cobrança total. Calcule a cobrança total de cada conta em `conta_dos_restaurantes` e dê ao array resultante o nome de `cobrancas_totais`.

In [ ]:
cobrancas_totais = ...
cobrancas_totais

Vamos ler alguns dados para usar na próxima pergunta.

In [ ]:
mais_contas_de_restaurantes = pd.read_csv("data/more_restaurant_bills.csv").get("Bill").values

**Pergunta 1.3.3.** O array `mais_contas_de_restaurantes` contém 100.000 contas! Calcule a cobrança total de cada uma, assumindo novamente uma gorjeta de vinte por cento, e dê ao array resultante o nome `mais_cobrancas_totais`.

In [ ]:
mais_cobrancas_totais = ...
mais_cobrancas_totais

A função `np.sum` leva um único array de números como argumento. Ele retorna a soma de todos os números desse array (portanto, retorna um único número, não um array).

**Pergunta 1.3.4.** Qual foi a soma de todas as contas em `mais_contas_de_restaurantes`, **incluindo gorjetas**?

In [ ]:
soma_das_contas = ...
soma_das_contas

##### Potências de Dois
As potências de 2 ($2^0 = 1$, $2^1 = 2$, $2^2 = 4$, etc) surgem frequentemente na ciência da computação. (Por exemplo, você deve ter notado que o armazenamento em smartphones ou computadores vem em potências de 2, como 64 GB, 128 GB ou 256 GB.)

**Pergunta 1.3.5.** Use `np.arange` e o operador de exponenciação `**` para criar um array contendo as primeiras 40 potências de 2, começando em $2^0=1$.

* **Dica 1:** Seu kernel “morreu” quando você executou sua solução? Existe uma resposta incorreta comum para esse problema que tenta criar um array com tantas entradas que o Python desiste e trava. Se isso acontecer com você, verifique sua resposta!

* **Dica 2:** Talvez comece com as primeiras 5 potências de dois. Depois de fazer isso funcionar, tente todos os 40. Em nenhum momento você deve escrever manualmente `0, 1, 2, 3, 4, ...`; se você estiver tentando isso, leia novamente algumas células anteriores desse notebook.

In [ ]:
potencias_de_2 = ...
potencias_de_2

# 2. DataFrames

## 2.1. Introdução

Para uma coleção de coisas no mundo, um array é útil para descrever um único atributo de cada coisa. Por exemplo, para a coleção de estados dos EUA, uma matriz poderia descrever a área territorial de cada um. As tabelas ampliam essa ideia descrevendo vários atributos para cada elemento de uma coleção. Numa tabela de estados, por exemplo, podemos registar a área territorial, a população, a capital do estado e o nome do governador. Em outras palavras, as tabelas rastreiam muitas entidades (indivíduos, armazenados como linhas) e, para cada entidade, muitos atributos (recursos, armazenados como colunas).

Na célula abaixo temos dois arrays. O primeiro contém a população mundial em cada ano (conforme estimado pelo US Census Bureau), e o segundo contém os próprios anos (em ordem, de modo que os primeiros elementos na população e as matrizes de anos correspondam).

In [ ]:
anos = np.arange(1950, 2022+1)
quantidade_populacional = pd.read_csv("data/world_population_2022.csv").get("Population").values

print("Coluna de população:", quantidade_populacional)
print("Coluna dos anos:", anos)

Suponha que queiramos responder a esta pergunta:

> Quando é que a população mundial ultrapassou os 7 mil milhões?

Você poderia tecnicamente responder a essa pergunta apenas olhando para as matrizes, mas é um pouco complicado, pois seria necessário contar a posição onde a população ultrapassou pela primeira vez os 7 bilhões e, em seguida, encontrar o elemento correspondente na matriz de anos. Em casos como estes, pode ser mais fácil colocar os dados em uma tabela.

Assim como `numpy` fornece arrays, um pacote popular chamado `pandas` fornece **DataFrames**, que é o nome em `pandas` para **tabelas**. `pandas` é *a* ferramenta para fazer ciência de dados em Python. Infelizmente, `pandas` não é tão fofo quanto o nome sugere: é muito complicado e pode ser um tanto difícil de aprender.

Em vez de usar `pandas`, usaremos um pacote que criamos especificamente para o curso do DSC10. É um *subconjunto* de `pandas`, incluindo apenas as partes que os criadores acharam necessárias e descartando todo o resto . Por ser menor (e mais fofo), eles deram para esse módulo o nome carinhoso de `babypandas`.

<img src='https://github.com/ThiagoPoppe/monitoria_fcd2024/blob/main/labs_gabs/lab01/imagens/pandas-babypandas.jpg?raw=true' width=400>

Você pode importar `babypandas` usando o seguinte código:

In [ ]:
import pandas as pd


O bom de `babypandas` é que é mais fácil de aprender *mas* cada pedaço de código que você escreve usando `babypandas` funcionará com `pandas` também. Se você é especialista em ciência de dados ou apenas fará muitas análises de dados em Python, verá muitos `pandas` no futuro.

A célula abaixo:

- cria um DataFrame vazio usando a expressão `pd.DataFrame()`,
- atribui duas colunas ao DataFrame chamando `assign`,
- atribui o DataFrame resultante ao nome `populacao_df` e, finalmente
- exibe `populacao_df` para que possamos ver o DataFrame que criamos.

`"Populacao"` e `"Ano"` são rótulos de coluna que escolhemos. Poderíamos ter escolhido qualquer coisa, mas é uma boa ideia escolher nomes que sejam descritivos e não muito longos.

In [ ]:
populacao_df = pd.DataFrame().assign(
    Populacao=quantidade_populacional,
    Ano=anos
)

populacao_df

Agora os dados estão todos juntos em um único DataFrame! É muito mais fácil analisar esses dados. Se você precisa saber qual era a população em 2011, por exemplo, você pode saber com um simples olhar. Revisitaremos este DataFrame mais tarde.

**Questão 2.1.1.** Na célula abaixo, criamos 2 arrays. Usando as etapas acima, atribua `top_10_filmes` a um DataFrame que possui duas colunas chamadas `Avaliacao` e `Nome`, que contêm `top_10_avaliacoes` e `top_10_nomes` respectivamente.

In [ ]:
top_10_avaliacoes = np.array([9.2, 9.2, 9., 8.9, 8.9, 8.9, 8.9, 8.9, 8.9, 8.8])
top_10_nomes = np.array([
        'The Shawshank Redemption (1994)',
        'The Godfather (1972)',
        'The Godfather: Part II (1974)',
        'Pulp Fiction (1994)',
        "Schindler's List (1993)",
        'The Lord of the Rings: The Return of the King (2003)',
        '12 Angry Men (1957)',
        'The Dark Knight (2008)',
        'Il buono, il brutto, il cattivo (1966)',
        'The Lord of the Rings: The Fellowship of the Ring (2001)'
])

In [ ]:
top_10_filmes = ...
top_10_filmes

Suponha que você queira adicionar suas próprias classificações a este DataFrame. A célula abaixo contém sua avaliação de cada filme:

In [ ]:
minhas_avaliacoes = [8, 2, 1, 9, 7, 10, 6, 4, 3, 5]

**Pergunta 2.1.2** Você também pode usar o método `assign` para adicionar uma coluna a um DataFrame já existente. Crie um novo DataFrame chamado `com_minhas_avaliacoes` adicionando uma coluna chamada `MinhaAvaliacao` ao DataFrame em `top_10_filmes`.

In [ ]:
com_minhas_avaliacoes = ...
com_minhas_avaliacoes

## 2.2. Índices

Você deve ter notado que o DataFrame de populações contém o que parece ser uma coluna extra e sem rótulo à esquerda com os números de 0 a 65. **Isto não é uma coluna, é o que chamamos de *índice***. O índice contém os rótulos das linhas. Enquanto as colunas deste DataFrame são rotuladas como `"Populacao"` e `"Ano"`, as linhas são rotuladas como 0, 1, ..., 65.

Por padrão, `babypandas` não sabe como rotular as linhas, então apenas as numera (começando com 0). É claro que, neste caso, faz mais sentido usar o ano como rótulo da linha. Podemos fazer isso dizendo ao `babypandas` para definir a coluna `"Ano"` como índice:

In [ ]:
populacao_por_ano = populacao_df.set_index('Ano')
populacao_por_ano

Como veremos, isso faz mais do que deixar o DataFrame mais bonito – é muito útil também.

**Pergunta 2.2.1** Crie um novo DataFrame chamado `top_10_filmes_por_nome` pegando o DataFrame que você criou acima, `top_10_filmes`, e definindo o índice como a coluna `Nome`.

In [ ]:
top_10_filmes_por_nome = ...
top_10_filmes_por_nome

Você pode obter um array de nomes de linhas usando `.index`. Por exemplo, o array de nomes de linhas do DataFrame `populacao_por_ano` é:

In [ ]:
populacao_por_ano.index

**Questão 2.2.2** Usando o código acima, atribua a `decimo_filme` o nome do décimo filme em `top_10_filmes_por_nome`.

* **Dica:** Lembre-se de que o índice é um array e usamos colchetes para acessar os elementos de um array. E também que o índice 0 corresponde ao 1º elemento do array.

In [ ]:
decimo_filme = ...
decimo_filme

## 2.3 Lendo um DataFrame de um arquivo
Na maioria dos casos, não teremos o trabalho de digitar todos os dados manualmente. Em vez disso, podemos usar funções fornecidas por `babypandas` para ler dados de arquivos externos.

A função `pd.read_csv()` pega um argumento, um caminho para um arquivo de dados (uma string) e retorna um DataFrame. Existem muitos formatos de arquivos de dados, mas CSV (*comma separated values*, ou "valores separados por vírgula") é o mais comum.

**Questão 2.3.1.** O arquivo `data/imdb.csv` contém informações sobre os 250 filmes mais bem avaliados no IMDb. Carregue-o como um DataFrame chamado `imdb`.

In [ ]:
imdb = ...
imdb

Observe os `...` no meio do DataFrame. Isso significa que muitas linhas foram omitidas. Este DataFrame é grande o suficiente para que apenas algumas de suas linhas sejam exibidas, mas as outras ainda estão lá. São 250 filmes no total.

De onde veio o `imdb.csv`? Se você entrar no diretório `data/`, deverá ver um arquivo chamado `imdb.csv`.

Abra o arquivo `imdb.csv` nessa pasta e observe o formato. O que você percebe? A terminação do nome do arquivo `.csv` indica que este arquivo está no formato [CSV (comma-separated value)](http://edoceo.com/utilitas/csv-file-format).

**Pergunta 2.3.2.** Este é um conjunto de dados de filmes, portanto faz sentido usar o título do filme como rótulo da linha. Crie um novo DataFrame chamado `imdb_por_nome` que usa o título do filme como índice.

In [ ]:
imdb_por_nome = ...
imdb_por_nome

## 2.4. Series



Suponha que estejamos interessados ​​principalmente nas classificações de filmes. Para extrair apenas esta coluna do DataFrame, usamos o método `.get`:

In [ ]:
avaliacoes = imdb_por_nome.get('Rating')
avaliacoes

Observe como não apenas as classificações do filme foram retornadas, mas também o nome do filme! Isto ocorre precisamente porque definimos o título do filme como o índice! Por exemplo, se tivéssemos solicitado a coluna `"Rating"` do DataFrame original, `imdb`, veríamos:

In [ ]:
imdb.get('Rating')

Esta é uma forma pela qual os índices são muito úteis – eles fornecem rótulos significativos para os dados.

À primeira vista, pode parecer que pedir uma coluna usando `.get` retorna um DataFrame com uma coluna, mas isso não está certo. Em vez disso, ele retorna um tipo especial de coisa chamado *Series*:

In [ ]:
type(imdb_por_nome.get('Rating'))

Você pode pensar em uma `Series` como um array com um índice. Enquanto as matrizes são sequências simples de números sem rótulos, `Series` pode ter rótulos. Isso geralmente é muito útil.

`avaliacoes` agora é uma `Series` que contém a coluna de classificações de filmes. Suponha que estejamos interessados ​​na avaliação de um filme específico: _Alien_. Para fazer isso, usaremos o "*acessador*" `.loc` que extrai um valor da Série em um *local* específico:

In [ ]:
avaliacoes.loc["Alien"]

Há algumas coisas a serem observadas aqui. Primeiro, esses são colchetes em torno de `"Alien"`. Isso ocorre porque `.loc` não é um método, mas um "*acessador*". Os colchetes sinalizam que iremos extrair um elemento da `Série`. Segundo, passamos o rótulo como uma string.

**Pergunta 2.4.1.** Encontre a avaliação de _3 Idiotas_ (*3 Idiots*).

In [ ]:
avaliacao_de_tres_idiotas = ...
avaliacao_de_tres_idiotas

Agora suponha que quiséssemos saber o ano em que _Alien_ foi lançado. Poderíamos fazer isso obtendo primeiro a coluna dos anos:

In [ ]:
anos = imdb_por_nome.get('Year')
anos

E então usando `.loc` para obter a entrada correta:

In [ ]:
anos.loc['Alien']

Também poderíamos fazer isso em uma única etapa *encadeando* as operações:

In [ ]:
imdb_por_nome.get('Year').loc['Alien']

Isso funciona porque o Python primeiro avalia `imdb_por_nome.get('Year')` como uma `Series`. Em seguida, avalia `.loc['Alien']` para retornar o ano.

O encadeamento é usado com bastante frequência e pode ser útil. Apenas certifique-se de não encadear muitas coisas que tornem seu código difícil de ler. Você sempre pode salvar um resultado intermediário em uma variável.

**Pergunta 2.4.2** Encontre a década em que _Gone Girl_ foi lançado usando encadeamento.

*Dica*: `imbd_por_nome` possui uma coluna chamada `"Decade"` (*Década*).

In [ ]:
decada = ...
decada

# 3. Analisando conjuntos de dados

Com apenas alguns métodos DataFrame, podemos responder algumas questões interessantes sobre o conjunto de dados IMDb.

Se quisermos apenas as avaliações dos filmes, podemos usar `.get`:

In [ ]:
avaliacoes = imdb_por_nome.get("Rating")
avaliacoes

Lembre-se de que `avaliacoes` é uma série. Objetos de série possuem alguns métodos úteis.

**Pergunta 3.1.** Encontre a maior avaliação no conjunto de dados.

*Dica:* Digite `avaliacoes.` e pressione Tab para ver uma lista dos métodos disponíveis. Existe algum que parece útil?

In [ ]:
maior_avaliacao = ...
maior_avaliacao

Você provavelmente quer saber o *nome* do filme de maior avaliação que encontrou anteriormente! Para fazer isso, podemos ordenar toda a série usando o método `.sort_values`:

In [ ]:
avaliacoes.sort_values()

Portanto, na verdade, existem dois filmes com maior audiência no conjunto de dados: *Um Sonho de Liberdade* e *O Poderoso Chefão*.

Observe que estamos ordenando pelas avaliações, e não pelos rótulos! Além disso, o rótulo segue a ordenação conforme a sua avaliação. Isto é exatamente o que queremos.

Quando utilizamos o método `sort_values`, a `Series` resultante tem os dados ordenados em ordem crescente, do menor ao maior. Este é o comportamento padrão de `sort_values`, mas podemos mudar isso. Se quiséssemos os filmes com melhor avaliação no topo, precisaríamos especificar que a avaliação não deveria ser em ordem crescente com um *argumento keyword* ("palavra-chave") opcional:


In [ ]:
avaliacoes.sort_values(ascending=False)

Se definirmos o argumento `ascending` como `True`, obteremos o mesmo resultado como se não o definissemos. Isso é o que queremos dizer quando dizemos que o comportamento padrão de `sort_values` é classificar em ordem crescente. Confirme se as próximas duas células fornecem a mesma saída.

In [ ]:
avaliacoes.sort_values(ascending=True)

In [ ]:
avaliacoes.sort_values()

Não só podemos ordenar séries, mas também ordenar DataFrames inteiros. Quando fazemos isso, temos que especificar a coluna pela qual iremos ordenar:

In [ ]:
imdb_por_nome.sort_values('Rating')

Da mesma forma, podemos especificar que a ordenação deve estar em ordem decrescente:

In [ ]:
imdb_por_nome.sort_values('Rating', ascending=False)

Alguns detalhes sobre a ordenação de um DataFrame:

1. O primeiro argumento para `sort_values` é o nome de uma coluna pela qual iremos ordenar.
2. Se a coluna contiver strings, `sort` ordenará em ordem alfabética; se a coluna tiver números, ela será ordenada numericamente.
3. `imdb_por_nome.sort_values("Rating")` retorna um novo DataFrame; o DataFrame `imdb_por_nome` não é modificado. Por exemplo, se chamarmos `imdb_por_nome.sort("Rating")`, então executar `imdb_por_nome` por si só ainda retornaria o DataFrame não ordenado. Para salvar o resultado, você deve atribuí-lo a uma nova variável.
4. As linhas sempre permanecem juntas quando um DataFrame é ordenado. Não faria sentido ordenar apenas uma coluna e deixar as outras colunas em paz. Por exemplo, neste caso, se ordenássemos apenas a coluna `"Rating"`, todos os filmes terminariam com avaliações erradas.

**Pergunta 3.2.** Crie uma versão de `imdb_por_nome` que seja ordenada cronologicamente, com os filmes mais antigos primeiro. Chame-o de `imdb_ordenado`.

In [ ]:
imdb_ordenado = ...
imdb_ordenado

**Pergunta 3.3.** Qual é o título do filme mais antigo no conjunto de dados? Você poderia simplesmente procurar isso na saída da célula anterior. Em vez disso, escreva o código Python para descobrir.

* **Dica:** Lembre-se de que o índice é um array.

In [ ]:
titulo_do_filme_mais_antigo = ...
titulo_do_filme_mais_antigo

Suponha que queiramos obter a avaliação do filme mais antigo no DataFrame. Uma maneira de fazer isso é primeiro encontrar o rótulo de índice do filme mais antigo (o que já fizemos). Em seguida, extraímos a coluna `"Rating"` e usamos `.loc` para encontrar a avaliação do filme mais antigo.

In [ ]:
imdb_ordenado.get('Rating').loc[titulo_do_filme_mais_antigo]

Porém, existe uma maneira mais rápida. Uma série não possui apenas um acessador `.loc`, mas também um acessador `.iloc`. Enquanto `.loc` procura coisas por *rótulo*, `.iloc` procura elementos por *posição inteira*.

Vamos lembrar o que está na coluna `"Rating"`:

In [ ]:
imdb_ordenado.get('Rating')

Se quisermos a avaliação da primeira linha, podemos usar `.iloc[0]`:

In [ ]:
imdb_ordenado.get('Rating').iloc[0]

Isso retorna exatamente a mesma coisa que `imdb_ordenado.get('Rating').loc['The Kid']`; essas são duas maneiras de fazer a mesma coisa. Normalmente é mais conveniente acessar um elemento por seu rótulo do que por sua posição inteira, mas é bom saber `.loc` e `.iloc`.

**Pergunta 3.4.** Qual é a avaliação do quinto filme mais antigo no conjunto de dados? Você poderia simplesmente procurar isso na saída da célula anterior. Em vez disso, escreva o código Python para descobrir.

In [ ]:
avaliacao_do_quinto_filme_mais_antigo = ...
avaliacao_do_quinto_filme_mais_antigo

# 4. Encontrar partes de um conjunto de dados

Suponha que você esteja interessado em filmes da década de 1950. Ordenar o DataFrame por ano não ajuda, porque a década de 1950 está no meio do conjunto de dados. Em vez disso, usaremos um recurso de Series que nos permite comparar facilmente cada elemento em uma coluna com um valor específico.

Primeiro lembre-se que podemos usar `.get` para extrair uma única coluna. O resultado não é um DataFrame, mas sim uma Series:

In [ ]:
imdb_por_nome.get('Decade')

Queremos verificar se cada filme foi lançado na década de 1950. Python nos dá uma maneira de verificar se duas coisas são iguais com `==` (lembre-se que `=` já está sendo usado para outro propósito: atribui valores a variáveis nomes):

In [ ]:
3 == 4

In [ ]:
3 == 3

`True` e `False` são instâncias de um tipo que não vimos antes:

In [ ]:
type(True)

`bool` significa "Boolean", em homenagem ao lógico inglês [George Boole](https://en.wikipedia.org/wiki/George_Boole). Dizemos que "True" e "False" são valores *Booleanos*.

Acontece que podemos facilmente verificar se *cada* um dos elementos em uma `Series` é igual a alguma coisa:

In [ ]:
imdb_por_nome.get('Decade') == 1950

Vemos que o resultado é uma nova série que tem `Verdadeiro` apenas onde a década foi 1950, e `Falso` em todos os outros lugares. Dizemos que a Series resultante é uma Series de *Booleanos*, ou uma *Series Booleana*.

Vamos chamar esse resultado de `eh_de_1950s`. Seu nome pode ser lido como se fosse uma pergunta: “esse filme é da década de 1950”?

In [ ]:
eh_de_1950s = imdb_por_nome.get('Decade') == 1950
eh_de_1950s

Cada linha é uma resposta a esta pergunta. *O Homem Elefante* é da década de 1950? `Falso`. *Tudo sobre Eva* é da década de 1950? `Verdadeiro`.

Podemos usar `eh_de_1950s` para selecionar apenas as linhas de `imdb_por_nome` para as quais a resposta é `Verdadeiro`. A sintaxe para isso é:

In [ ]:
imdb_por_nome[eh_de_1950s]

O que `imdb_por_nome[eh_de_1950s]` faz, precisamente, é percorrer `imdb_por_nome` linha por linha. Se a linha chamada *Singin' in the Rain* tiver o valor `True` em `eh_de_1950s`, essa linha será mantida. Se o valor for `False`, a linha será descartada. E assim por diante, para cada linha.

Observe que poderíamos ter conseguido isso sem nunca criar a variável `eh_de_1950s`, simplesmente colocando o código que usamos para criar a Series booleana diretamente dentro de `[...]`. Este é um padrão típico que você usará muito!

In [ ]:
imdb_por_nome[imdb_por_nome.get('Decade') == 1950]

Ajuda ler os colchetes como "onde". Portanto, o comando na célula acima diz para manter todas as linhas de `imdb_por_nome` *onde* a década é a década de 1950.

Criar um novo DataFrame selecionando apenas certas linhas de um DataFrame existente que satisfaça alguma condição é chamado de *consulta*. A linha de código `imdb_por_nome[imdb_por_nome.get('Decade') == 1950]` é uma *consulta*.

**Questão 4.1.** Crie um DataFrame chamado `noventa_e_oito` contendo os filmes lançados em 1998.

In [ ]:
noventa_e_oito = ...
noventa_e_oito

Até agora só descobrimos onde uma coluna é *exatamente* igual a um determinado valor. No entanto, existem muitos outros operadores de comparação que poderíamos usar. Aqui estão alguns:

|Operador|Testes|
|-|-|
|`==`|a coisa da esquerda é igual à coisa da direita|
|`!=`|a coisa à esquerda *não* é igual à coisa à direita|
|`>`|a coisa à esquerda é maior que (e não igual) à coisa à direita|
|`>=`|a coisa à esquerda é maior ou igual à coisa à direita|
|`<`|a coisa à esquerda é menor que (e não igual) à coisa à direita|

As [notas de curso](https://notes.dsc10.com/02-data_sets/querying.html#examples) do DSC10 tem mais exemplos.

**Questão 4.2.** Utilizando os operadores da tabela acima, encontre todos os filmes com avaliação superior a 8,6. Coloque seus dados em um DataFrame chamado `realmente_bem_avaliados`.

In [ ]:
realmente_bem_avaliados = ...
realmente_bem_avaliados

Qual é a maior avaliação de qualquer filme da década de 1990? Agora temos as ferramentas para responder a perguntas como essas. Dividindo em pedaços, encontramos primeiro todos os filmes da década de 1990:

In [ ]:
eh_de_1990s = imdb_por_nome.get('Decade') == 1990
eh_de_1990s

Em seguida, selecionamos apenas estes filmes em nosso DataFrame:

In [ ]:
de_1990s = imdb_por_nome[eh_de_1990s]
de_1990s

Encontramos então a maior avaliação apenas destes filmes:

In [ ]:
de_1990s.get('Rating').max()

Ou, se quiséssemos fazer tudo isso de forma mais concisa usando encadeamento:

In [ ]:
imdb_por_nome[imdb_por_nome.get('Decade') == 1990].get('Rating').max()

**Pergunta 4.3.** Encontre a avaliação média para filmes lançados no século 20 e a avaliação média para filmes lançados no século 21 para os filmes no `imdb`.

*Dica*: As séries possuem um método `.mean()`. Observe que o ano 2000 está no século 20 e que o filme mais antigo do conjunto de dados é de 1921!

In [ ]:
avaliacao_media_do_seculo_20 = ...
avaliacao_media_do_seculo_20

In [ ]:
avaliacao_media_do_seculo_21 = ...
avaliacao_media_do_seculo_21

A propriedade `shape` informa quantas linhas e colunas existem em um DataFrame. (Uma "propriedade" é similar a um método que não precisa ser chamado adicionando parênteses.)

In [ ]:
imdb_por_nome.shape

Como um array, você pode obter o primeiro elemento do `shape` usando `[0]` e o segundo elemento usando `[1]`. Por exemplo, o número de linhas em `imdb_por_nome` é:

In [ ]:
imdb_por_nome.shape[0]

Podemos usar isso para responder "Quantos filmes são do século 20?":

In [ ]:
imdb_por_nome[imdb_por_nome.get('Year') <= 2000].shape[0]

**Pergunta 4.4.** Use `shape` (e aritmética) para encontrar a *proporção* de filmes no conjunto de dados que foram lançados no século 20 e a proporção do século 21.

* **Dica:** A *proporção* de filmes lançados no século 20 é o *número* de filmes lançados no século 20, dividido pelo *número total* de filmes no conjunto de dados.

In [ ]:
proporcao_do_seculo_20 = ...
proporcao_do_seculo_20

In [ ]:
proporcao_do_seculo_21 = ...
proporcao_do_seculo_21

**Pergunta 4.5.** Finalmente, vamos revisitar o DataFrame `populacao_por_ano` do início do laboratório. Calcule o ano em que a população mundial ultrapassou pela primeira vez os 7 mil milhões.

In [ ]:
ano_que_a_populacao_ultrapassou_7_bilhoes = ...
ano_que_a_populacao_ultrapassou_7_bilhoes

# Linha de chegada

Parabéns! Você concluiu o Laboratório 1.

Para enviar sua tarefa:

1. Selecione `Kernel -> Reiniciar e executar tudo` para garantir que você executou todas as células.
2. Leia o caderno para ter certeza de que está tudo bem.
4. Baixe seu notebook usando `Arquivo -> Baixar como -> Notebook (.ipynb)` e, em seguida, carregue seu notebook para o Moodle.